File: assets/js/matrix.js
/**
 * ALGEBRIFY - DYNAMIC MATRIX CALCULATOR
 * Calls Python/Flask backend (/api/matrix) for all matrix mathematical operations:
 * addition, subtraction, scalar mult, matrix mult, transpose, determinant,
 * inverse, rank, trace, and step-by-step breakdown.
 */

// ==========================================================================
// FRACTION & NUMBER FORMATTING UTILS
// ==========================================================================

function formatNumber(val, tolerance = 1e-6) {
    if (isNaN(val) || !isFinite(val)) return "Undefined";
    if (Math.abs(val) < tolerance) return "0";

    // Check if close to integer
    if (Math.abs(val - Math.round(val)) < tolerance) {
        return Math.round(val).toString();
    }

    // Try finding a clean fraction for denominators up to 100
    for (let denom = 2; denom <= 100; denom++) {
        const num = Math.round(val * denom);
        if (Math.abs(val - num / denom) < tolerance) {
            return `\\frac{${num}}{${denom}}`;
        }
    }

    // Otherwise round to 4 decimals cleanly without trailing zeros
    const rounded = parseFloat(val.toFixed(4));
    return rounded.toString();
}

function formatPlainNumber(val, tolerance = 1e-6) {
    if (isNaN(val) || !isFinite(val)) return "Undefined";
    if (Math.abs(val) < tolerance) return "0";
    if (Math.abs(val - Math.round(val)) < tolerance) return Math.round(val).toString();
    return parseFloat(val.toFixed(4)).toString();
}

function parseInputNumber(str) {
    if (typeof str === "number") return isNaN(str) ? 0 : str;
    if (!str || typeof str !== "string") return 0;
    str = str.trim();
    if (!str) return 0;
    if (str.includes("/")) {
        const parts = str.split("/");
        if (parts.length === 2) {
            const num = parseFloat(parts[0]);
            const den = parseFloat(parts[1]);
            if (!isNaN(num) && !isNaN(den) && Math.abs(den) > 1e-9) {
                return num / den;
            }
        }
    }
    const val = parseFloat(str);
    return isNaN(val) ? 0 : val;
}

// ==========================================================================
// DYNAMIC MATRIX CALCULATOR CLASS
// ==========================================================================

class DynamicMatrixCalculator {
    constructor() {
        this.rowsA = 2;
        this.colsA = 2;
        this.rowsB = 2;
        this.colsB = 2;

        this.matrixAContainer = document.getElementById("matrix-a-grid");
        this.matrixBContainer = document.getElementById("matrix-b-grid");
        this.resultContainer = document.getElementById("matrix-calc-result");

        this.init();
    }

    init() {
        this.renderGrids();
        this.bindDimensionControls();
        this.bindOperationButtons();
        this.populateSampleValues();
    }

    bindDimensionControls() {
        const rowsASelect = document.getElementById("rows-a");
        const colsASelect = document.getElementById("cols-a");
        const rowsBSelect = document.getElementById("rows-b");
        const colsBSelect = document.getElementById("cols-b");

        if (rowsASelect && colsASelect) {
            rowsASelect.addEventListener("change", (e) => {
                const oldVals = this.getMatrixA();
                this.rowsA = parseInt(e.target.value, 10);
                this.renderGridA(oldVals);
            });
            colsASelect.addEventListener("change", (e) => {
                const oldVals = this.getMatrixA();
                this.colsA = parseInt(e.target.value, 10);
                this.renderGridA(oldVals);
            });
        }

        if (rowsBSelect && colsBSelect) {
            rowsBSelect.addEventListener("change", (e) => {
                const oldVals = this.getMatrixB();
                this.rowsB = parseInt(e.target.value, 10);
                this.renderGridB(oldVals);
            });
            colsBSelect.addEventListener("change", (e) => {
                const oldVals = this.getMatrixB();
                this.colsB = parseInt(e.target.value, 10);
                this.renderGridB(oldVals);
            });
        }

        // Preset Dimension Buttons
        const presetBtns = document.querySelectorAll(".matrix-preset-btn");
        presetBtns.forEach(btn => {
            btn.addEventListener("click", () => {
                presetBtns.forEach(b => b.classList.remove("active"));
                btn.classList.add("active");

                const rA = parseInt(btn.getAttribute("data-ra"), 10) || 2;
                const cA = parseInt(btn.getAttribute("data-ca"), 10) || 2;
                const rB = parseInt(btn.getAttribute("data-rb"), 10) || rA;
                const cB = parseInt(btn.getAttribute("data-cb"), 10) || cA;

                this.rowsA = rA;
                this.colsA = cA;
                this.rowsB = rB;
                this.colsB = cB;

                if (rowsASelect) rowsASelect.value = rA;
                if (colsASelect) colsASelect.value = cA;
                if (rowsBSelect) rowsBSelect.value = rB;
                if (colsBSelect) colsBSelect.value = cB;

                this.renderGrids();
                this.populateSampleValues();
            });
        });
    }

    renderGrids() {
        this.renderGridA();
        this.renderGridB();
    }

    renderGridA(preserveVals = null) {
        if (!this.matrixAContainer) return;
        this.matrixAContainer.style.gridTemplateColumns = `repeat(${this.colsA}, 1fr)`;
        this.matrixAContainer.innerHTML = "";

        for (let r = 0; r < this.rowsA; r++) {
            for (let c = 0; c < this.colsA; c++) {
                const input = document.createElement("input");
                input.type = "number";
                input.step = "any";
                input.id = `a_${r}_${c}`;
                input.placeholder = "0";
                if (preserveVals && preserveVals[r] && preserveVals[r][c] !== undefined) {
                    input.value = preserveVals[r][c];
                }
                input.addEventListener("focus", () => input.select());
                input.addEventListener("click", () => input.select());
                input.addEventListener("keydown", (e) => {
                    if (e.key === "Enter") this.add();
                });
                this.matrixAContainer.appendChild(input);
            }
        }
    }

    renderGridB(preserveVals = null) {
        if (!this.matrixBContainer) return;
        this.matrixBContainer.style.gridTemplateColumns = `repeat(${this.colsB}, 1fr)`;
        this.matrixBContainer.innerHTML = "";

        for (let r = 0; r < this.rowsB; r++) {
            for (let c = 0; c < this.colsB; c++) {
                const input = document.createElement("input");
                input.type = "number";
                input.step = "any";
                input.id = `b_${r}_${c}`;
                input.placeholder = "0";
                if (preserveVals && preserveVals[r] && preserveVals[r][c] !== undefined) {
                    input.value = preserveVals[r][c];
                }
                input.addEventListener("focus", () => input.select());
                input.addEventListener("click", () => input.select());
                input.addEventListener("keydown", (e) => {
                    if (e.key === "Enter") this.add();
                });
                this.matrixBContainer.appendChild(input);
            }
        }
    }

    populateSampleValues() {
        const sampleA = [
            [1, 2, 3],
            [4, 5, 6],
            [7, 8, 9]
        ];
        const sampleB = [
            [9, 8, 7],
            [6, 5, 4],
            [3, 2, 1]
        ];

        for (let r = 0; r < this.rowsA; r++) {
            for (let c = 0; c < this.colsA; c++) {
                const input = document.getElementById(`a_${r}_${c}`);
                if (input && (!input.value || input.value === "0")) {
                    if (sampleA[r] && sampleA[r][c] !== undefined) {
                        input.value = sampleA[r][c];
                    } else {
                        input.value = r === c ? "1" : "0";
                    }
                }
            }
        }

        for (let r = 0; r < this.rowsB; r++) {
            for (let c = 0; c < this.colsB; c++) {
                const input = document.getElementById(`b_${r}_${c}`);
                if (input && (!input.value || input.value === "0")) {
                    if (sampleB[r] && sampleB[r][c] !== undefined) {
                        input.value = sampleB[r][c];
                    } else {
                        input.value = r === c ? "1" : "0";
                    }
                }
            }
        }
    }

    getMatrixA() {
        const matrix = [];
        for (let r = 0; r < this.rowsA; r++) {
            const row = [];
            for (let c = 0; c < this.colsA; c++) {
                const el = document.getElementById(`a_${r}_${c}`);
                const val = parseInputNumber(el?.value);
                row.push(val);
            }
            matrix.push(row);
        }
        return matrix;
    }

    getMatrixB() {
        const matrix = [];
        for (let r = 0; r < this.rowsB; r++) {
            const row = [];
            for (let c = 0; c < this.colsB; c++) {
                const el = document.getElementById(`b_${r}_${c}`);
                const val = parseInputNumber(el?.value);
                row.push(val);
            }
            matrix.push(row);
        }
        return matrix;
    }

    matrixToLatex(matrix) {
        if (!matrix || matrix.length === 0) return "\\begin{bmatrix}\\end{bmatrix}";
        const rows = matrix.map(row => row.map(val => formatNumber(val)).join(" & "));
        return `\\begin{bmatrix} ${rows.join(" \\\\ ")} \\end{bmatrix}`;
    }

    showResult(htmlContent) {
        if (!this.resultContainer) return;
        this.resultContainer.innerHTML = htmlContent;
        if (typeof renderAllMath === "function") {
            renderAllMath(this.resultContainer);
        }
    }

    showError(message) {
        this.showResult(`
            <div class="alert-message alert-error">
                <strong>⚠️ Dimension / Compatibility Error:</strong> ${message}
            </div>
        `);
    }

    bindOperationButtons() {
        const addBtn = document.getElementById("op-add-btn");
        const subBtn = document.getElementById("op-sub-btn");
        const mulBtn = document.getElementById("op-mul-btn");
        const scalarBtn = document.getElementById("op-scalar-btn");
        const transABtn = document.getElementById("op-trans-a-btn");
        const detABtn = document.getElementById("op-det-a-btn");
        const invABtn = document.getElementById("op-inv-a-btn");
        const rankABtn = document.getElementById("op-rank-a-btn");
        const traceABtn = document.getElementById("op-trace-a-btn");
        const clearBtn = document.getElementById("op-clear-btn");

        if (addBtn) addBtn.addEventListener("click", () => this.add());
        if (subBtn) subBtn.addEventListener("click", () => this.subtract());
        if (mulBtn) mulBtn.addEventListener("click", () => this.multiply());
        if (scalarBtn) scalarBtn.addEventListener("click", () => this.scalarMultiply());
        if (transABtn) transABtn.addEventListener("click", () => this.transposeA());
        if (detABtn) detABtn.addEventListener("click", () => this.determinantA());
        if (invABtn) invABtn.addEventListener("click", () => this.inverseA());
        if (rankABtn) rankABtn.addEventListener("click", () => this.rankA());
        if (traceABtn) traceABtn.addEventListener("click", () => this.traceA());
        if (clearBtn) clearBtn.addEventListener("click", () => this.clear());
    }

    // ====== OPERATIONS DELEGATED TO FLASK BACKEND ======

    async add() {
        const A = this.getMatrixA();
        const B = this.getMatrixB();

        const response = await algebrifyApi("/api/matrix", {
            operation: "add",
            matrixA: A,
            matrixB: B
        });

        if (!response.success) {
            this.showError(response.error || "Addition failed.");
            return;
        }

        this.showResult(`
            <div class="math-display">
                \\[ A + B = ${response.latex_a} + ${response.latex_b} = ${response.latex_result} \\]
            </div>
            <div class="calc-steps">
                <strong>Step-by-Step Addition (Python Flask Backend):</strong>
                <p>${response.explanation}</p>
            </div>
        `);
    }

    async subtract() {
        const A = this.getMatrixA();
        const B = this.getMatrixB();

        const response = await algebrifyApi("/api/matrix", {
            operation: "subtract",
            matrixA: A,
            matrixB: B
        });

        if (!response.success) {
            this.showError(response.error || "Subtraction failed.");
            return;
        }

        this.showResult(`
            <div class="math-display">
                \\[ A - B = ${response.latex_a} - ${response.latex_b} = ${response.latex_result} \\]
            </div>
            <div class="calc-steps">
                <strong>Step-by-Step Subtraction (Python Flask Backend):</strong>
                <p>${response.explanation}</p>
            </div>
        `);
    }

    async multiply() {
        const A = this.getMatrixA();
        const B = this.getMatrixB();

        const response = await algebrifyApi("/api/matrix", {
            operation: "multiply",
            matrixA: A,
            matrixB: B
        });

        if (!response.success) {
            this.showError(response.error || "Multiplication failed.");
            return;
        }

        const steps = response.steps || [];
        this.showResult(`
            <div class="math-display">
                \\[ A \\cdot B = ${response.latex_a} \\cdot ${response.latex_b} = ${response.latex_result} \\]
            </div>
            <div class="calc-steps">
                <strong>Step-by-Step Dot Products (Python Flask Backend):</strong>
                <ol>
                    ${steps.slice(0, 8).map(step => `<li>${step}</li>`).join("")}
                    ${steps.length > 8 ? `<li>... and so on for all ${steps.length} entries.</li>` : ""}
                </ol>
            </div>
        `);
    }

    async scalarMultiply() {
        const scalarInput = document.getElementById("scalar-k");
        const k = scalarInput && scalarInput.value !== "" ? parseFloat(scalarInput.value) : 2;
        const A = this.getMatrixA();

        const response = await algebrifyApi("/api/matrix", {
            operation: "scalar_multiply",
            matrixA: A,
            scalar: k
        });

        if (!response.success) {
            this.showError(response.error || "Scalar multiplication failed.");
            return;
        }

        this.showResult(`
            <div class="math-display">
                \\[ ${k} \\cdot A = ${k} \\cdot ${response.latex_a} = ${response.latex_result} \\]
            </div>
            <div class="calc-steps">
                <strong>Step-by-Step Scalar Multiplication (Python Flask Backend):</strong>
                <p>${response.explanation}</p>
            </div>
        `);
    }

    async transposeA() {
        const A = this.getMatrixA();

        const response = await algebrifyApi("/api/matrix", {
            operation: "transpose",
            matrixA: A
        });

        if (!response.success) {
            this.showError(response.error || "Transpose failed.");
            return;
        }

        this.showResult(`
            <div class="math-display">
                \\[ A^T = \\left( ${response.latex_a} \\right)^T = ${response.latex_result} \\]
            </div>
            <div class="calc-steps">
                <strong>Transpose Property (Python Flask Backend):</strong>
                <p>Swapped rows and columns. Dimensions transformed from ${this.rowsA}×${this.colsA} to ${this.colsA}×${this.rowsA}.</p>
            </div>
        `);
    }

    async determinantA() {
        if (this.rowsA !== this.colsA) {
            this.showError(`Determinant is only defined for square matrices ($n \\times n$). Matrix A is ${this.rowsA}×${this.colsA}.`);
            return;
        }

        const A = this.getMatrixA();
        const response = await algebrifyApi("/api/matrix", {
            operation: "determinant",
            matrixA: A
        });

        if (!response.success) {
            this.showError(response.error || "Determinant failed.");
            return;
        }

        this.showResult(`
            <div class="math-display">
                \\[ \\det(A) = ${response.vmatrix_latex} = ${response.formatted_det} \\]
            </div>
            <div class="calc-steps">
                <strong>Interpretation (Python Flask Backend):</strong>
                <p>${response.interpretation}</p>
            </div>
        `);
    }

    async inverseA() {
        if (this.rowsA !== this.colsA) {
            this.showError(`Inverse is only defined for square matrices ($n \\times n$). Matrix A is ${this.rowsA}×${this.colsA}.`);
            return;
        }

        const A = this.getMatrixA();
        const response = await algebrifyApi("/api/matrix", {
            operation: "inverse",
            matrixA: A
        });

        if (!response.success) {
            this.showError(response.error || "Could not compute inverse for this matrix.");
            return;
        }

        this.showResult(`
            <div class="math-display">
                \\[ A^{-1} = ${response.latex_result} \\]
            </div>
            <div class="calc-steps">
                <strong>Verification (Python Flask Backend):</strong>
                <p>${response.explanation}</p>
            </div>
        `);
    }

    async rankA() {
        const A = this.getMatrixA();
        const response = await algebrifyApi("/api/matrix", {
            operation: "rank",
            matrixA: A
        });

        if (!response.success) {
            this.showError(response.error || "Rank computation failed.");
            return;
        }

        this.showResult(`
            <div class="math-display">
                \\[ \\text{Rank}(A) = ${response.rank} \\]
            </div>
            <div class="calc-steps">
                <strong>Rank Explanation (Python Flask Backend):</strong>
                <p>${response.explanation}</p>
            </div>
        `);
    }

    async traceA() {
        if (this.rowsA !== this.colsA) {
            this.showError(`Trace is only defined for square matrices ($n \\times n$). Matrix A is ${this.rowsA}×${this.colsA}.`);
            return;
        }

        const A = this.getMatrixA();
        const response = await algebrifyApi("/api/matrix", {
            operation: "trace",
            matrixA: A
        });

        if (!response.success) {
            this.showError(response.error || "Trace computation failed.");
            return;
        }

        this.showResult(`
            <div class="math-display">
                \\[ \\text{tr}(A) = \\sum_{i=1}^{${this.rowsA}} a_{ii} = ${response.diagonal_terms} = ${response.formatted_trace} \\]
            </div>
            <div class="calc-steps">
                <strong>Trace Property (Python Flask Backend):</strong>
                <p>${response.explanation}</p>
            </div>
        `);
    }

    clear() {
        for (let r = 0; r < this.rowsA; r++) {
            for (let c = 0; c < this.colsA; c++) {
                const el = document.getElementById(`a_${r}_${c}`);
                if (el) el.value = "";
            }
        }
        for (let r = 0; r < this.rowsB; r++) {
            for (let c = 0; c < this.colsB; c++) {
                const el = document.getElementById(`b_${r}_${c}`);
                if (el) el.value = "";
            }
        }
        this.showResult("<p class='text-center text-light'>Cleared all inputs. Select an operation to calculate.</p>");
    }
}

// Auto-instantiate if on a page with matrix calculator
document.addEventListener("DOMContentLoaded", () => {
    if (document.getElementById("matrix-a-grid")) {
        window.matrixCalculatorInstance = new DynamicMatrixCalculator();
    }
});

File: backend/matrix.py
"""
ALGEBRIFY BACKEND - MATRIX ENGINE
Implements core matrix algebra: Addition, Subtraction, Multiplication,
Scalar Multiplication, Transpose, Determinant, Inverse, Rank, and Trace.
Includes detailed mathematical step-by-step explanations and LaTeX outputs.
"""

import numpy as np
import sympy as sp

try:
    from .utils import format_number, format_plain_number, matrix_to_latex
except (ImportError, ValueError):
    from utils import format_number, format_plain_number, matrix_to_latex


def matrix_add(matrix_a, matrix_b):
    """
    Compute Matrix Addition: C = A + B
    """
    A = np.array(matrix_a, dtype=float)
    B = np.array(matrix_b, dtype=float)

    if A.shape != B.shape:
        return {
            "success": False,
            "error": f"Addition requires identical dimensions. Matrix A is {A.shape[0]}×{A.shape[1]} but Matrix B is {B.shape[0]}×{B.shape[1]}."
        }

    C = A + B
    res_list = C.tolist()

    return {
        "success": True,
        "operation": "add",
        "result": res_list,
        "latex_a": matrix_to_latex(matrix_a),
        "latex_b": matrix_to_latex(matrix_b),
        "latex_result": matrix_to_latex(res_list),
        "explanation": "Add corresponding entries: $C_{ij} = A_{ij} + B_{ij}$."
    }


def matrix_subtract(matrix_a, matrix_b):
    """
    Compute Matrix Subtraction: C = A - B
    """
    A = np.array(matrix_a, dtype=float)
    B = np.array(matrix_b, dtype=float)

    if A.shape != B.shape:
        return {
            "success": False,
            "error": f"Subtraction requires identical dimensions. Matrix A is {A.shape[0]}×{A.shape[1]} but Matrix B is {B.shape[0]}×{B.shape[1]}."
        }

    C = A - B
    res_list = C.tolist()

    return {
        "success": True,
        "operation": "subtract",
        "result": res_list,
        "latex_a": matrix_to_latex(matrix_a),
        "latex_b": matrix_to_latex(matrix_b),
        "latex_result": matrix_to_latex(res_list),
        "explanation": "Subtract corresponding entries: $C_{ij} = A_{ij} - B_{ij}$."
    }


def matrix_multiply(matrix_a, matrix_b):
    """
    Compute Matrix Multiplication: C = A * B
    """
    A = np.array(matrix_a, dtype=float)
    B = np.array(matrix_b, dtype=float)

    rows_a, cols_a = A.shape
    rows_b, cols_b = B.shape

    if cols_a != rows_b:
        return {
            "success": False,
            "error": f"Matrix multiplication $A \\times B$ is only defined when columns of $A$ equal rows of $B$. Here, columns of $A$ = {cols_a} and rows of $B$ = {rows_b}."
        }

    C = np.matmul(A, B)
    res_list = C.tolist()

    # Generate step-by-step dot product breakdown
    steps = []
    for i in range(rows_a):
        for j in range(cols_b):
            terms = [f"({format_plain_number(A[i, k])} \\cdot {format_plain_number(B[k, j])})" for k in range(cols_a)]
            steps.append(f"$c_{{{i+1}{j+1}}} = {' + '.join(terms)} = {format_number(C[i, j])}$")

    return {
        "success": True,
        "operation": "multiply",
        "result": res_list,
        "latex_a": matrix_to_latex(matrix_a),
        "latex_b": matrix_to_latex(matrix_b),
        "latex_result": matrix_to_latex(res_list),
        "steps": steps,
        "explanation": "Compute the dot product of each row in $A$ with each column in $B$."
    }


def matrix_scalar_multiply(matrix_a, scalar_k):
    """
    Compute Scalar Multiplication: C = k * A
    """
    A = np.array(matrix_a, dtype=float)
    try:
        k = float(scalar_k)
    except (ValueError, TypeError):
        k = 1.0

    C = k * A
    res_list = C.tolist()

    return {
        "success": True,
        "operation": "scalar_multiply",
        "scalar": k,
        "result": res_list,
        "latex_a": matrix_to_latex(matrix_a),
        "latex_result": matrix_to_latex(res_list),
        "explanation": f"Multiply every entry of Matrix A by $k = {format_plain_number(k)}$."
    }


def matrix_transpose(matrix_a):
    """
    Compute Matrix Transpose: A^T
    """
    A = np.array(matrix_a, dtype=float)
    AT = A.T
    res_list = AT.tolist()

    return {
        "success": True,
        "operation": "transpose",
        "result": res_list,
        "latex_a": matrix_to_latex(matrix_a),
        "latex_result": matrix_to_latex(res_list),
        "explanation": "Swap rows and columns: $(A^T)_{ij} = A_{ji}$."
    }


def matrix_determinant(matrix_a):
    """
    Compute Matrix Determinant: det(A)
    """
    A = np.array(matrix_a, dtype=float)
    rows, cols = A.shape

    if rows != cols:
        return {
            "success": False,
            "error": f"Determinant is only defined for square matrices. Matrix A is {rows}×{cols}."
        }

    det_val = float(np.linalg.det(A))
    if abs(det_val) < 1e-9:
        det_val = 0.0

    vmatrix_rows = [" & ".join(format_plain_number(v) for v in row) for row in matrix_a]
    vmatrix_latex = f"\\begin{{vmatrix}} {' \\\\ '.join(vmatrix_rows)} \\end{{vmatrix}}"

    is_invertible = abs(det_val) > 1e-9
    interpretation = (
        f"✨ Non-zero determinant ($\\det(A) = {format_number(det_val)} \\neq 0$): Matrix $A$ is <strong>invertible (non-singular)</strong> and has full rank ({rows})."
        if is_invertible
        else f"⚠️ Zero determinant ($\\det(A) = 0$): Matrix $A$ is <strong>singular (non-invertible)</strong> with linearly dependent columns."
    )

    return {
        "success": True,
        "operation": "determinant",
        "determinant": det_val,
        "formatted_det": format_number(det_val),
        "vmatrix_latex": vmatrix_latex,
        "is_invertible": is_invertible,
        "interpretation": interpretation
    }


def matrix_inverse(matrix_a):
    """
    Compute Matrix Multiplicative Inverse: A^-1
    """
    A = np.array(matrix_a, dtype=float)
    rows, cols = A.shape

    if rows != cols:
        return {
            "success": False,
            "error": f"Inverse is only defined for square matrices. Matrix A is {rows}×{cols}."
        }

    det_val = float(np.linalg.det(A))
    if abs(det_val) < 1e-9:
        return {
            "success": False,
            "error": "Matrix is singular (det(A) = 0) and has no multiplicative inverse."
        }

    try:
        sp_mat = sp.Matrix(matrix_a)
        inv_mat = sp_mat.inv()
        inv_list = [[float(v.evalf()) for v in inv_mat.row(r)] for r in range(rows)]
        
        return {
            "success": True,
            "operation": "inverse",
            "determinant": det_val,
            "result": inv_list,
            "latex_a": matrix_to_latex(matrix_a),
            "latex_result": matrix_to_latex(inv_list),
            "explanation": f"Since $\\det(A) = {format_number(det_val)} \\neq 0$, the inverse satisfies $A \\cdot A^{{-1}} = I_{{{rows}}}$."
        }
    except Exception as e:
        return {
            "success": False,
            "error": f"Could not invert matrix: {str(e)}"
        }


def matrix_rank(matrix_a):
    """
    Compute Matrix Rank: Rank(A)
    """
    A = np.array(matrix_a, dtype=float)
    rank_val = int(np.linalg.matrix_rank(A))
    rows, cols = A.shape

    is_full_rank = (rank_val == min(rows, cols))
    explanation = (
        f"Matrix $A$ has <strong>Full Rank</strong> ($\\text{{Rank}}(A) = {rank_val}$). All {rank_val} rows/columns are linearly independent."
        if is_full_rank
        else f"Matrix $A$ is <strong>Rank-Deficient</strong> ($\\text{{Rank}}(A) = {rank_val} < \\min({rows}, {cols})$). It contains linearly dependent rows/columns."
    )

    return {
        "success": True,
        "operation": "rank",
        "rank": rank_val,
        "rows": rows,
        "cols": cols,
        "is_full_rank": is_full_rank,
        "latex_a": matrix_to_latex(matrix_a),
        "explanation": explanation
    }


def matrix_trace(matrix_a):
    """
    Compute Matrix Trace: tr(A) = sum(A_ii)
    """
    A = np.array(matrix_a, dtype=float)
    rows, cols = A.shape

    if rows != cols:
        return {
            "success": False,
            "error": f"Trace is only defined for square matrices. Matrix A is {rows}×{cols}."
        }

    diag_vals = [A[i, i] for i in range(rows)]
    trace_val = float(sum(diag_vals))
    diag_terms = " + ".join(format_plain_number(v) for v in diag_vals)

    return {
        "success": True,
        "operation": "trace",
        "trace": trace_val,
        "formatted_trace": format_number(trace_val),
        "latex_a": matrix_to_latex(matrix_a),
        "diagonal_terms": diag_terms,
        "explanation": f"Sum of main diagonal elements: $\\text{{tr}}(A) = {diag_terms} = {format_number(trace_val)}$."
    }
